MessengerProj/src/main/java/com/b44t/messenger/LaunchActivity.java

```diff
@@ -753,7 +755,16 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
                 else {
                     switch (permissions[i]) {
                         case Manifest.permission.READ_CONTACTS:
-                            msg += "- " + ApplicationLoader.applicationContext.getString(R.string.PermissionContacts) + "\n\n";
+                            SharedPreferences preferences = ApplicationLoader.applicationContext.getSharedPreferences("general", Context.MODE_PRIVATE);
+                            if (preferences.getBoolean("PermissionContactsDialogShown", false)) {
+                                // The dialog was already shown once
+                                grantedCount++;
+                            } else {
+                                msg += "- " + ApplicationLoader.applicationContext.getString(R.string.PermissionContacts) + "\n\n";
+                                SharedPreferences.Editor editor = preferences.edit();
+                                editor.putBoolean("PermissionContactsDialogShown", true);
+                                editor.apply();
+                            }
                             break;
                         case Manifest.permission.WRITE_EXTERNAL_STORAGE:
```

Comment:
WRITE_EXTERNAL_STOREAGE and RECORD_AUDIO seems not be a problem as they're queried only on explicit user actions?

Reply:
Actually, WRITE_EXTERNAL_STORAGE is also queried on every startup, and only RECORD_AUDIO is not (I didn't realize this).

Reply:
Yes, but querying the WRITE_EXTERNAL_STORAGE permission comes after a user action, so I think this is always fine to ask here. Same as for RECORD_AUDIO - if the user _clicks actively_ on the microphone and the app has not the permission to record audio, it is fine to ask the user, even a thousand times - what else can the app do?